ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc

```diff
@@ -0,0 +1,547 @@
+ *
+ *   custom:
+ *     Each entries are separated by ','
+ *     Each entries have property_key=value format.
+ *     There must be no speces.
+ *
+ *     Supported props:
+ *       input_rank: (mandatory)
+ *            Rank of each input tensors.
+ *            Each ranks are separeted by ':'.
+ *            The number of ranks must be the same as the number of input
+ *            tensors.
+ *       enable_tensorrt: (optional)
+ *            set true to enable NVIDIA TensorRT. GPU acceleration must be
+ *            enabled in order to use it.
+ *
+ *     Examples:
+ *       tensor_filter framework=mxnet model=model/Inception-BN.json
+ *                input=1:3:224:224
+ *                inputname=data
+ *                accelerator=true:!cpu,!npu,gpu
+ *                ...
+ *                custom=input_rank=4,enable_tensorrt=true
+ *
+ *        tensor_filter framework=mxnet model=model/Inception-BN.json
+ *                input=1:3:224:224,1
+ *                inputname=data1,data2
+ *                ...
+ *                custom=input_rank=4:1
+ */
+#include <nnstreamer_cppplugin_api_filter.hh>
+#include <nnstreamer_util.h>
+#include <stdexcept>
+#include <tensor_common.h>
+#include <unistd.h>
+
+#include <fcntl.h>
+#include <glib.h>
+#include <glib/gstdio.h>
+#include <sys/stat.h>
+#include <sys/types.h>
+
+#include <mxnet-cpp/MxNetCpp.h>
+
+using nnstreamer::tensor_filter_subplugin;
+using namespace mxnet::cpp;
+
+namespace nnstreamer
+{
+namespace tensorfilter_mxnet
+{
+
+const static std::string kFileLocation = "/ext/nnstreamer/tensor_filter/tensor_filter_mxnet.cc";
+const static std::string kFileUrl
+    = "https://github.com/nnstreamer/nnstreamer/tree/main" + kFileLocation;
+
+extern "C" {
+void init_filter_mxnet (void) __attribute__ ((constructor)); /**< Dynamic library contstructor */
+void fini_filter_mxnet (void) __attribute__ ((destructor)); /**< Dynamic library desctructor */
+}
+
+class TensorFilterMXNet final : public tensor_filter_subplugin
+{
+  public:
+  static void init_filter (); /**< Dynamic library contstructor helper */
+  static void fini_filter (); /**< Dynamic library desctructor helper */
+
+  TensorFilterMXNet ();
+  ~TensorFilterMXNet ();
+
+  /**< Implementations of tensor_filter_subplugin */
+  tensor_filter_subplugin &getEmptyInstance ();
+  void configure_instance (const GstTensorFilterProperties *prop);
+  void invoke (const GstTensorMemory *input, GstTensorMemory *output);
+  void getFrameworkInfo (GstTensorFilterFrameworkInfo &info);
+  int getModelInfo (model_info_ops ops, GstTensorsInfo &in_info, GstTensorsInfo &out_info);
+  int eventHandler (event_ops ops, GstTensorFilterFrameworkEventData &data);
+
+  static const std::string ext_symbol; /**< extension of model symbol (.json) */
+  static const std::string ext_params; /**< extension of model parameters (.params) */
+
+  /**< Data type for MXNet NDArray from <mshadow/base.h> which causes errors when included */
+  enum TypeFlag : int {
```

Comment:
Does including ```https://github.com/apache/incubator-mxnet/blob/master/include/mxnet/lib_api.h```  also make compile errors?

Reply:
In my opinion, the original author seems to want to support all data types NNStreamer supports defined here: https://github.com/nnstreamer/nnstreamer/blob/main/gst/nnstreamer/include/tensor_typedef.h#L153

But I think `lib_api.h` you referred is more appropriate choice and data types mxnet supports listed [here](https://mxnet.apache.org/versions/1.7.0/api/python/docs/api/ndarray/ndarray.html#mxnet.ndarray.Cast) is matched with the `lib_api.h`. Therefore I can change the TypeFlag enum to use items listed in `lib_api.h` only and remove items which are not listed there. Then some data types will be unsupported but it will be safer choice.